WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,777 @@
+      }
+      else
+      {
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+         /* calculate the actual return offset */
+         usOffset += ((u32_t)(pComrvStackFrame->uiCalleeToken) - usOffset) & (--usOverlayGroupSize);
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+      }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+   }
+   /* we are calling a new overlay function or accessing overlay data */
+   else
+   {
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+     /* Update comrv stack frame with the loaded multi group table entry.
+        It is used to calculate the actual return offset in case we
+        are returning to a multi group token */
+      pComrvStackFrame->usCalleeMultiGroupTableEntry = usSelectedMultiGroupEntry;
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+   }
+
+#ifdef D_COMRV_FW_INSTRUMENTATION
+   comrvNotificationHook(uiProfilingIndication , unToken.ucValue);
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+
+   /* group is now loaded to memory so we can return the address of the data/function */
+   return (void*)((u08_t*)pAddress + usOffset);
+}
+
+/**
+* Get comrv heap eviction candidates according to a given size
+*
+* @param requestedEvictionSize - size requested for eviction; expressed in
+*                                units of D_COMRV_OVL_GROUP_SIZE_MIN
+*        pEvictCandidatesList  - output eviction candidate list of comrv heap indexes
+* @return number of eviction candidates in the output list
+*/
+u08_t comrvGetEvictionCandidates(u08_t ucRequestedEvictionSize, u08_t* pEvictCandidatesList)
+{
+   u08_t ucAccumulatedSize = 0, ucIndex = 0;
+   u08_t ucEntryIndex, ucNumberOfCandidates = 0;
+   u32_t uiEvictCandidateMap[D_COMRV_EVICT_CANDIDATE_MAP_SIZE], uiCandidates;
+
+   /* first lets clear the uiCandidates list */
+   memset(uiEvictCandidateMap, 0, sizeof(u32_t)*D_COMRV_EVICT_CANDIDATE_MAP_SIZE);
+
+#ifdef D_COMRV_EVICTION_LRU
+   /* get the first lru entry */
+   ucEntryIndex = stComrvCB.ucLruIndex;
+   /* loop LRU list until the eviction uiCandidates accumulated size
+      reaches the requested eviction size */
+   do
+   {
+      /* verify the entry isn't locked */
+      if (stComrvCB.stOverlayCache[ucEntryIndex].unProperties.stFields.ucLocked != D_COMRV_ENTRY_LOCKED)
+      {
+         /* count the number of uiCandidates */
+         ucNumberOfCandidates++;
+         /* accumulate size */
+         ucAccumulatedSize += stComrvCB.stOverlayCache[ucEntryIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits;
+         /* set the eviction candidate in the eviction map */
+         uiEvictCandidateMap[ucEntryIndex/D_COMRV_DWORD_IN_BITS] |= (1 << ucEntryIndex);
+      }
+      /* move to the next LRU candidate */
+      ucEntryIndex = stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex;
+   /* loop as long as we didn't get to the requested eviction size or we reached end of the list
+      (means that all entries are locked) */
+   } while (ucAccumulatedSize < ucRequestedEvictionSize && ucEntryIndex != D_COMRV_LRU_LAST_ITEM);
+
```

Comment:
add a condition check if accumulated size < requested size, add an error "not enough entries"

Reply:
done in 5e17026